### PR TITLE
Js-frogger

### PR DESCRIPTION
**Pull request by maxovsanyuk**
### Classic Arcade Game Clone Project

[Demo](https://maxovsanyuk.github.io/frogger/) | 
[Code base](https://github.com/maxovsanyuk/frontend-2019-homeworks/blob/js-frogger/submissions/maxovsanyuk/js-frogger/js/app.js)

he code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+
+
+// Now instantiate your objects.
+// Place all enemy objects in an array called allEnemies
+// Place the player object in a variable called player
+const enemy1 = new Enemy(0,40);
+const enemy2 = new Enemy(0,120);
+const enemy3 = new Enemy(0,200);
+const enemy4 = new Enemy(-200,120);
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: 1. Here and above, too many "magic numbers". Please, employ constants named so they explain the numbers behind. You also definitely know that arrays also can be constants.

2. What if you would be required to create a hundred of enemies? Please, employ some better scalable approach. You've already got an array to keep all enemies together in the code below. Is there any good reason to keep these fragments apart?

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    }
+};
+
+// Update the enemy's position, required method for game
+// Parameter: dt, a time delta between ticks
+Enemy.prototype.update = function(dt) {
+    this.x += this.speed * dt;
+    this.y
+
+    this.handlCollision(player);
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: `player` must be a global variable. This violates encapsulation OOP principle. The enemy must explicitly know the subject to collide with.

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+Enemy.prototype.handlCollision = function(player){
+    if( player.x < this.x + 80 &&
+        player.x + 70 > this.x &&
+        player.y < this.y + 40 &&
+        player.y + 40 > this.y){
+        player.x = 200;
+        player.y = 360;
+        player.health--
+        if(player.health === 0){
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Is it an enemy who reset the player and decides if the game is over? Hardly so. The only function of an enemy is to move an take health from the player. 

Please, take other logic out of enemy to the more appropriate place.

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+};
+
+
+
+// Now write your own player class
+// This class requires an update(), render() and
+// a handleInput() method.
+
+
+var Player = function(x,y) {
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Player and Enemy classes have something in common. Actually, pretty many things.
Please, find the ways to make your code DRYer.

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 3 above)
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: You might combine code from lines 26 and 27 in a smart way. You may need refreshing yourself on the difference between `i--` and `--i`. Please, employ your native JS knowledge.

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 1 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
const player = new Player();
const ROW_HEIGHT = 80;
const ROW_DIFF = 40;
const POSITION_DIRECTION_Y =[1,1,2,2,3,].map(function(heightY){
    return (heightY * ROW_HEIGHT) - ROW_DIFF ;
});

const allEnemies = POSITION_DIRECTION_Y.map(function(instance){
    return new Enemy(instance,player);
})
```

Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```
const player = new Player();
const ROW_HEIGHT = 80;
const ROW_DIFF = 40;
const POSITION_DIRECTION_Y =[1,1,2,2,3,].map(function(heightY){
    return (heightY * ROW_HEIGHT) - ROW_DIFF ;
});

const allEnemies = POSITION_DIRECTION_Y.map(function(instance){
    return new Enemy(instance,player);
})
```

Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
const player = new Player();
const ROW_HEIGHT = 80;
const ROW_DIFF = 40;
const POSITION_DIRECTION_Y =[1,1,2,2,3,].map(function(heightY){
    return (heightY * ROW_HEIGHT) - ROW_DIFF ;
});

const allEnemies = POSITION_DIRECTION_Y.map(function(instance){
    return new Enemy(instance,player);
})
```

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 3 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
 this.player.initialPosition();
            --this.player.health;
            Player.prototype.gameOver();    
```

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            Player.prototype.gameOver();            
+         }
+};
+
+// Update the enemy's position, required method for game
+// Parameter: dt, a time delta between ticks
+Enemy.prototype.update = function(dt) {
+    this.x += this.speed * dt;
+
+    this.handlCollision(player);
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Still refers to a global variable here. Obviously just missed to refactor here as other location refer to the player instance properly.

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    }
+     this.score = INITIAL_PLAYER_DATA.score,
+     this.health = INITIAL_PLAYER_DATA.health
+};
+
+Player.prototype.reachFinish = function(){
+    if (this.y < 0) {
+        setTimeout(() => {
+            this.initialPosition();
+            player.score++
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Why referring to a global var not to just `this`?

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        }, 300);
+    }
+}
+
+Player.prototype.gameOver = function(){
+    if(player.health === 0){
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Global var reference. counter-OOP.
Please, eliminate referring to global variables within classes everywhere.

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+Enemy.prototype.handlCollision = function(player){
+    if( this.x < player.x + 80 &&
+        this.x + 70 > player.x &&
+        this.y < player.y + 40 &&
+        this.y + 40 > player.y)
+        {
+            this.player.initialPosition();
+            --this.player.health;
+            Player.prototype.gameOver();
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Why not `this.player.gameOver()`?

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    // we've provided one for you to get started
+
+    // The image/sprite for our enemies, this uses
+    // a helper we've provided to easily load images
+    this.sprite = "images/char-boy.png";
+    this.initialPosition();
+    this.initialData();
+};
+
+Player.prototype.initialPosition = function() {
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Method names should be or start with verbs. E.g. `goToInitialPosition` (as per method body context).

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    const INITIAL_PLAYER_POSITION ={
+      x: 200,
+      y: 360  
+    }
+      this.x = INITIAL_PLAYER_POSITION.x,
+      this.y = INITIAL_PLAYER_POSITION.y
+};
+
+Player.prototype.initialData = function() {
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: Method names should be or start with verbs. E.g. `resetState` or `initializeState`

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    const INITIAL_PLAYER_DATA ={
+     score: 0,
+     health: 2  
+    }
+     this.score = INITIAL_PLAYER_DATA.score,
+     this.health = INITIAL_PLAYER_DATA.health
+};
+
+Player.prototype.reachFinish = function(){
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: `reachFinish` reads rather like **go to finish** than **test or check if finish is reached**
Please, give appropriate name.

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+Player.prototype.handleInput = function(key){
+    if (key === 'left' && this.x >= 50){
+        this.x -= 101;
+    }else if(key === 'right' && this.x <= 400){
+        this.x += 101;
+    }else if(key === 'up' && this.y >= 0){
+        this.y -= 80;
+        this.reachFinish();
+    }else if(key === 'down' && this.y <= 300){
+        this.y += 80;
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: There is a constant for this. Please, use it.
Global constants are OK.
Please, also introduce constants for magic numbers like 50, 400, 300 above and elsewhere within the code base.

File: submissions/maxovsanyuk/js-frogger/js/app.js [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    this.player = player;
+    this.sprite = 'images/enemy-bug.png';
+    this.randomSpeed();
+};
+
+Enemy.prototype.randomSpeed = function(){
+    this.speed = ( Math.random() + 1 ) * 120 ;
+  };
+
+Enemy.prototype.handlCollision = function(player){
```
Comment (OleksiyRudenko) [inline submissions/maxovsanyuk/js-frogger/js/app.js]: A typo: handlCollision `=>` handl**e**Collision

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 8 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
            this.player.gameOver();            
```

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 10 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
Player.prototype.resetState = function() {
```

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 12 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
           const FIELD_LEFT_END = 50;
    const FIELD_RIGHT_END = 400;
    const FIELD_TOP_END = 0;
    const FIELD_LEFT_BOTTOM = 300;
    const STEP_SIZE_X = 101;
    const STEP_SIZE_Y = 80;
```

**Comment by OleksiyRudenko** [comment]
@maxovsanyuk Please, click **Resolve conversation** for items you have resolved.
This will let the code reviewer understanding if the codebase is ready for next review iteration.
Thanks.

**Comment by maxovsanyuk** [comment]
Ok, sure)Thanks

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 6 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
    if (this.y < 0) {
        setTimeout(() => {
            this.goToInitialPosition();
            this.score++
```

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 13 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
Enemy.prototype.handleCollision = function (player) {
```

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 7 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
Player.prototype.gameOver = function () {
    if (this.health === 0) {
```

File: submissions/maxovsanyuk/js-frogger/js/app.js (hunk 5 above)
Comment (maxovsanyuk) [reply submissions/maxovsanyuk/js-frogger/js/app.js]:
```suggestion
this.handleCollision(this.player);
```

**Pull request merged**